lldb/tools/lldb-dap/DAP.cpp

```diff
@@ -1310,103 +1396,161 @@ void DAP::EventThread() {
               continue;
 
             llvm::StringRef module_id = p_module->id;
 
-            const bool module_exists = modules.contains(module_id);
+            const bool module_exists =
+                dap_instance->modules.contains(module_id);
             if (remove_module && module_exists) {
-              modules.erase(module_id);
-              Send(protocol::Event{
+              dap_instance->modules.erase(module_id);
+              dap_instance->Send(protocol::Event{
                   "module", ModuleEventBody{std::move(p_module).value(),
                                             ModuleEventBody::eReasonRemoved}});
             } else if (module_exists) {
-              Send(protocol::Event{
+              dap_instance->Send(protocol::Event{
                   "module", ModuleEventBody{std::move(p_module).value(),
                                             ModuleEventBody::eReasonChanged}});
             } else if (!remove_module) {
-              modules.insert(module_id);
-              Send(protocol::Event{
+              dap_instance->modules.insert(module_id);
+              dap_instance->Send(protocol::Event{
                   "module", ModuleEventBody{std::move(p_module).value(),
                                             ModuleEventBody::eReasonNew}});
             }
           }
+        } else if (event_mask & lldb::SBTarget::eBroadcastBitNewTargetSpawned) {
+          auto target = lldb::SBTarget::GetTargetFromEvent(event);
+          auto target_index = debugger.GetIndexOfTarget(target);
+
+          // Set the shared debugger for GPU processes
+          DAPSessionManager::GetInstance().SetSharedDebugger(target_index,
+                                                             debugger);
+
+          // We create "attachCommands" that will select the target that already
+          // exists in LLDB. The DAP instance will attach to this already
+          // existing target and the debug session will be ready to go.
+          llvm::json::Object attach_config;
+          llvm::json::Array attach_commands;
+
+          attach_commands.push_back(llvm::formatv("target list").str());
+          attach_commands.push_back(
+              llvm::formatv("target select {0}", target_index).str());
+
+          // If we have a process name, add command to attach to the same
+          // process name
+
+          attach_config.try_emplace("type", "lldb");
+          attach_config.try_emplace("name", "GPU Session");
+          attach_config.try_emplace("attachCommands",
+                                    std::move(attach_commands));
+          attach_config.try_emplace("targetIdx", target_index);
+
+          // 2. Construct the main 'startDebugging' request arguments.
+          llvm::json::Object start_debugging_args;
+          start_debugging_args.try_emplace("request", "attach");
+          start_debugging_args.try_emplace("configuration",
+                                           std::move(attach_config));
```

Comment:
This might simplify this code pattern

```suggestion
          llvm::json::Object start_debugging_args {
             {"request", "attach"},
             {"configuration", std::move(attach_config)}
          };

```